csi/csi.go

```
@@ -257,18 +236,108 @@ func (s *OsdCsiServer) addEncryptionInfoToLabels(labels, csiSecrets map[string]s
+	// create correlation interceptor
+	var unaryInterceptors []grpc.UnaryServerInterceptor
+	correlationInterceptor := correlation.ContextInterceptor{
+		Origin: correlation.ComponentCSIDriver,
+	}
+	opts := make([]grpc.ServerOption, 0)
+	unaryInterceptors = append(unaryInterceptors, correlationInterceptor.ContextUnaryServerInterceptor)
+
+	// create scheduler interceptor
+	switch config.SchedulerName {
+	case "kubernetes":
+		logrus.Infof("CSI K8s filter being added for %s scheduler", config.SchedulerName)
+		ki := k8s.NewInterceptor()
+		unaryInterceptors = append(unaryInterceptors, ki.SchedUnaryInterceptor)
+
+	default:
+		logrus.Infof("No CSI filter being added for %s scheduler", config.SchedulerName)
+	}
+
+	// Add interceptors
+	opts = append(opts, grpc.UnaryInterceptor(grpc_middleware.ChainUnaryServer(unaryInterceptors...)))
+
+	// Create server
+	gServer, err := grpcserver.New(&grpcserver.GrpcServerConfig{
+		Name:    "CSI 1.7",
+		Net:     config.Net,
+		Address: config.Address,
+		Opts:    opts,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create CSI server: %v", err)
+	}
+
+	return gServer, nil
+}
+
+func autoSocketRecover(s *OsdCsiServer, stopCh chan struct{}) error {
+	socketPath := s.Address()
+	ticker := time.NewTicker(csiSocketCheckInterval)
+
+	// Start checking for CSI socket delete
+	for {
+		select {
+		case <-stopCh:
+			return nil
+		case <-ticker.C:
+		}
+
+		// Check if socket deleted
+		_, err := os.Stat(socketPath)
+		if err == nil {
+			continue
+		}
+
+		logrus.Infof("Detected CSI socket deleted at path %s. Stopping CSI gRPC server", socketPath)
+		s.GrpcServer.Stop()
+
+		// Re-create gRPC server
+		gServer, err := createGrpcServer(s.config)
+		if err != nil {
+			return fmt.Errorf("failed to re-create gRPC server: %v", err)
+		}
+		s.GrpcServer = gServer
+
+		// Start server
+		logrus.Infof("Restarting CSI gRPC server at %s", socketPath)
+		if err := s.Start(); err != nil {
```

Instead of failing fast, can we continuously retry?